integrations/outlook/src/client.ts

```diff
@@ -0,0 +1,116 @@
+  message: Message
+  ack: AckFunction
+  body: any
+}
+
+interface LifecycleEvent {
+  subscriptionId: string
+  subscriptionExpirationDateTime: string
+  tenantId: string
+  clientState: string
+  lifecycleEvent: 'subscriptionRemoved' | 'missed' | 'reauthorizationRequired'
+}
+
+export class GraphApi {
+  private client: Client
+  constructor(tenantId: string, clientId: string, clientSecret: string) {
+    const credential = new ClientSecretCredential(tenantId, clientId, clientSecret)
+    const authProvider = new TokenCredentialAuthenticationProvider(credential, {
+      scopes: ['https://graph.microsoft.com/.default'],
+    })
+
+    this.client = Client.initWithMiddleware({
+      //debugLogging: true,
+      authProvider,
+    })
+  }
+
+  /**
+   * Send Mail
+   */
+  public sendMail = async ({ message, conversation, ctx, ack, body }: SendMessageProps): Promise<void> => {
+    const refMessageId = conversation.tags['outlook:refMessageId']
+
+    if (!refMessageId) {
+      console.log('conv tag missing: outlook:refMessageId')
+      return
+    }
+
+    try {
+      await this.client
+        .api(`/users/${ctx.configuration.emailAddress}/messages/${refMessageId}/replyAll`)
+        .responseType(ResponseType.RAW)
+        .post({
+          message: {
+            body,
+          },
+        })
+      ack({ tags: { 'outlook:id': `${message.id}` } })
+    } catch (error) {
+      console.log(error.message)
+    }
+
+    return
+  }
+
+  /**
+   * Subscribe webhook
+   */
+  public subscribeWebhook = async (webhookUrl: string, ctx: IntegrationContext): Promise<string> => {
+    const expirationDateTime = this.generateExpirationDate()
+
+    const res = await this.client.api('/subscriptions').post({
+      changeType: 'created',
+      notificationUrl: webhookUrl,
+      lifecycleNotificationUrl: webhookUrl,
+      expirationDateTime,
+      resource: `/users/${ctx.configuration.emailAddress}/mailFolders('${ctx.configuration.mailFolder}')/messages`,
+    })
+    return res.id
+  }
+
+  /**
+   * Handle subscription life cycle events
+   */
+  public handleLifecycleEvents = async (event: LifecycleEvent) => {
+    if (event.lifecycleEvent === 'reauthorizationRequired') {
+      console.log('lifecycleEvent - reauthorizationRequired')
+
+      const expirationDateTime = this.generateExpirationDate()
+      await this.client.api(`/subscriptions/${event.subscriptionId}`).patch({
+        expirationDateTime,
+      })
+      console.log('webhook reauthorization success')
+    }
+
+    return
+  }
+  /**
+   * Unsubscribe Webhook
+   */
+  public unsubscribeWebhook = async (subscriptionId: string) => {
+    const res = await this.client.api(`/subscriptions/${subscriptionId}`).del()
+
+    return res
+  }
+
+  public getNotificationContent = async (odataId: string) => {
+    const res = await this.client.api(odataId).get()
+
+    return res
+  }
+
+  private generateExpirationDate = () => {
+    return moment.utc().add(20, 'minutes')
```

Comment:
ok I think I get it. Basically, Outlook will call the webhook with an event to notify that the webhook is about to be expired ? Then you call the `handleLifecycleEvents` method ?

If so, is there any risk of race conditions ? Will we miss some messages in between webhook expiry ? Does outlook call **before** the webhook is expired or when it already is ?

If this is not safe enough, we can use Botpress recurring events to refresh the webhook every 5 min (keeping the expiry at 20 min)

anyway, let me know if I understand correctly and let's discuss this

Reply:
1- "Basically, Outlook will call the webhook with an event to notify that the webhook is about to be expired ? Then you call the handleLifecycleEvents method ?":
Yes, exactly. Outlook will send the lifecycle event **before** the webhook expires to prevent downtime.

2-"Will we miss some messages in between webhook expiry ?":
The webhook may continue for a while, giving you extra time to respond. However, eventually change notification delivery pauses, until you take the required action. Any notifications about resource changes that happen when the change notification delivery pauses and the time when the app successfully creates the subscription again would be lost.

3-"is there any risk of race conditions ?":
I think there is no risk of the race condition, Outlook sends a notification at a time and expects a successful response.
other than that, there was a situation in my local dev env where the "regiter" method of the integration was executed before unregistering the previous subscription and caused to receive duplicated messages.


extra notes: maximum time for the webhook to expire is  4,230 minutes (under 3 days)

Reply:
I see no problem with keeping 20 minutes then! Let's keep it like that.

Reply:
ok, we're also implementing a way to check if there's an existing active subscripcion to avoid duplicated messages